src/compiler/program.ts

```diff
@@ -219,7 +219,12 @@ module ts {
             // Create the emit resolver outside of the "emitTime" tracking code below.  That way
             // any cost associated with it (like type checking) are appropriate associated with
             // the type-checking counter.
-            let emitResolver = getDiagnosticsProducingTypeChecker().getEmitResolver(sourceFile);
+            //
+            // If the -out option is specified, we should not pass the source file to getEmitResolver.
+            // This is because in the -out scenario all files need to be emitted, and therefore all
+            // files need to be type checked. And the way to specify that all files need to be type
+            // checked is to not pass the file to getEmitResolver.
+            let emitResolver = getDiagnosticsProducingTypeChecker().getEmitResolver(options.out ? undefined : sourceFile);

```

Comment:
nice catch.  I'm wondering if there's a way to make this less error prone (though nothing is coming to mind).  To some extent, what is weird is a model where a source file is passed at all to emit when you have -out.  I mean, what does it mean to say "emit foo.ts" or "emit bar.ts" when you have "--out"?  There's no effective difference, and it probably makes more sense for this to not be supported, or for the caller to have to say "emit all".  (i'm just speaking out loud.  i don't have an actual proposal).